README
TimePeriod - Python module for determining if a datetime is within a time
period. This is a Python version of Perl's Time::Period module.

Author: Paul Boyd <boyd.paul2@gmail.com>

Installation
----------------------------------------

To install from an archive:

$ tar zxf TimePeriod-X.Y.tar.gz
$ cd TimePeriod-X.Y
$ sudo ./setup.py install

Usage
----------------------------------------

You can use this module like this:

>>> from TimePeriod import inPeriod
>>> inPeriod ("wd {th}")
1

If you want to use a time other than the current time, pass a date time after the period:

m datetime import datetime
>>> from TimePeriod import inPeriod
>>> inPeriod ("wd {th}", datetime (2007, 1, 1))
0

Now we know with certainty that New Year's day 2007 was not on Thursday.

The TimePeriod module uses the same format as Perl's Time::Period module, which is detailed here:

http://search.cpan.org/~pryan/Period-1.20/Period.pm

Differences from Time::Period
----------------------------------------

There are a few intended differences from Time::Period:

1) An exception will be raised for an invalid format. Time::Period returns -1.
2) TimePeriod uses Python's datetime object, instead of an epoch.
3) Years are considered to have a maximum of 366 days, not 365.

License
----------------------------------------

Released under the LGPL:

http://www.gnu.org/licenses/#LGPL

Also see the LICENSE file.


CHANGES
Version 1.02:
    * Bug: "none" matching nothing (formerly it matched everything).
    * Bug: Inverted year ranges (e.g. 2015-2010) are now equivalent to standard
      formatted ranges.
    * Bug: "12pm" and "12am" are handled correctly for hours now.
    * Bug: Empty ranges (e.g. "wk {}") are no longer an error, but they never
      match.
    * Bug: The week number calculation is fixed:
        - Previously, the days 1-7 would be the first week, 8-14 would be the
          second week, etc.
        - Now, it counts the number of Sunday's that have passed.

Version 1.01:
    * Minor doc updates

Version 1.00:
    * Initial release


test.py
#!/usr/bin/env python
"""
Format tests
-------------------

>>> inPeriod ("yr{2007}", datetime (2007, 1, 1))
1

>>> inPeriod ("yr {2007}", datetime (2007, 1, 1))
1

>>> inPeriod ("yr{ 2007 }", datetime (2007, 1, 1))
1

>>> inPeriod ("yr { 2007 }", datetime (2007, 1, 1))
1

>>> inPeriod ("yr { 2007 - 2008 }", datetime (2007, 1, 1))
1

>>> inPeriod ("yr{ 2007-2008 }", datetime (2007, 1, 1))
1

>>> inPeriod ("yr{2007-2008}", datetime (2007, 1, 1))
1

>>> inPeriod ("yr {2007 2008}", datetime (2007, 1, 1))
1

Two options
>>> inPeriod ("yr {2007} wd {mon}", datetime (2007, 1, 1))
1

Two options with a comma
>>> inPeriod ("yr {2007}, min {5}", datetime (2007, 6, 12, 12, 5))
1

One option from each scale has to be true
>>> inPeriod ("yr {2007} wd {mon}", datetime (2007, 1, 2))
0

>>> inPeriod ("yr {2007} wd {tue-mon}", datetime (2007, 1, 1))
1

Options seperated by a comma are ORed together
>>> inPeriod ("min {5}, yr {2007}", datetime (2007, 6, 12))
1

Constants
>>> inPeriod ("always")
1

>>> inPeriod ("")
1

>>> inPeriod ("never")
0

>>> inPeriod ("min {5} min {10} min {15}", datetime (2007, 6, 12, 12, 15))
1

>>> inPeriod ("nv {2007} wd {tue-mon}", datetime (2007, 1, 1))
Traceback (most recent call last):
InvalidFormat: nv is not a valid scale.

>>> inPeriod ("x {2007}", datetime (2007, 1, 1))
Traceback (most recent call last):
InvalidFormat: x is not a valid scale.

>>> inPeriod ("none", datetime (2007, 1, 1))
0

Year tests
-------------------

>>> inPeriod ("yr {2007}", datetime (2007, 6, 1))
1

>>> inPeriod ("yr {2007-2009}", datetime (2009, 6, 1))
1

>>> inPeriod ("yr {2007 2009}", datetime (2007, 6, 1))
1

>>> inPeriod ("yr {2007 2009}", datetime (2008, 6, 1))
0

Numbers less than 100 assume the century of the datetime
>>> inPeriod ("yr {7}", datetime (2007, 6, 1))
1

>>> inPeriod ("yr {99}", datetime (1999, 6, 1))
1

>>> inPeriod ("yr {99}", datetime (2099, 6, 1))
1

>>> inPeriod ("yr {fred}", datetime (2099, 6, 1))
Traceback (most recent call last):
InvalidFormat: An integer value is required for year.

Long name
>>> inPeriod ("year {2009}", datetime (2009, 6, 12))
1

>>> inPeriod('yr {2015-2010}', datetime(2011, 1, 1))
1

>>> inPeriod('yr {1960}', datetime(1960, 1, 1))
1

Month tests
-------------------

>>> inPeriod ("mo {jun}", datetime (2099, 6, 1))
1

>>> inPeriod ("mo {jun-jul}", datetime (2099, 7, 1))
1

>>> inPeriod ("mo {nov-feb}", datetime (2099, 1, 1))
1

>>> inPeriod ("mo {jan mar jun}", datetime (2099, 3, 1))
1

>>> inPeriod ("mo {1-3}", datetime (2099, 2, 1))
1

Case insensitive and only the first three letters matter
>>> inPeriod ("mo {JUN-julio}", datetime (2099, 7, 1))
1

>>> inPeriod ("mo {13}", datetime (2099, 7, 1))
Traceback (most recent call last):
InvalidFormat: 13 is not valid for month. Valid options are between 1 and 12.

>>> inPeriod ("mo {0}", datetime (2099, 7, 1))
Traceback (most recent call last):
InvalidFormat: 0 is not valid for month. Valid options are between 1 and 12.

>>> inPeriod ("mo {garbledygook}", datetime (2099, 7, 1))
Traceback (most recent call last):
InvalidFormat: An integer value is required for month.

>>> inPeriod ("month {june}", datetime (2007, 6, 12))
1

Week tests
-------------------

>>> inPeriod ("wk {3}", datetime (2007, 6, 15))
1

>>> inPeriod ("wk {1 3 5}", datetime (2007, 6, 30))
1

>>> inPeriod ("wk {3-5}", datetime (2007, 6, 23))
1

>>> inPeriod ("wk {3-5}", datetime (2007, 6, 2))
0

>>> inPeriod ("wk {7}", datetime (2007, 6, 15))
Traceback (most recent call last):
InvalidFormat: 7 is not valid for week. Valid options are between 1 and 6.

>>> inPeriod ("wk {0}", datetime (2007, 6, 15))
Traceback (most recent call last):
InvalidFormat: 0 is not valid for week. Valid options are between 1 and 6.

>>> inPeriod ("wk {xxx}", datetime (2007, 6, 15))
Traceback (most recent call last):
InvalidFormat: An integer value is required for week.

>>> inPeriod ("week {3}", datetime (2007, 6, 12))
1

>>> inPeriod('wk {5-1}', datetime(2010, 12, 20))
0

>>> inPeriod('wk {6}', datetime(2011, 1, 31))
1

>>> inPeriod('wk {2}', datetime(2011, 1, 2))
1

>>> inPeriod('wk {1}', datetime(2011, 5, 1))
1

Year day tests
-------------------

>>> inPeriod ("yd {162}", datetime (2007, 6, 11))
1

>>> inPeriod ("yd {162}", datetime (2007, 6, 12))
0

>>> inPeriod ("yd {160-165}", datetime (2007, 6, 11))
1

>>> inPeriod ("yd {160 162 164}", datetime (2007, 6, 11))
1

Leap years have 366 days
>>> inPeriod ("yd {366}", datetime (2008, 12, 31))
1

>>> inPeriod ("yd {367}", datetime (2007, 6, 11))
Traceback (most recent call last):
InvalidFormat: 367 is not valid for year day. Valid options are between 1 and 366.

>>> inPeriod ("yd {0}", datetime (2007, 6, 11))
Traceback (most recent call last):
InvalidFormat: 0 is not valid for year day. Valid options are between 1 and 366.

>>> inPeriod ("yd {george}", datetime (2007, 6, 11))
Traceback (most recent call last):
InvalidFormat: An integer value is required for year day.

>>> inPeriod ("yday {163}", datetime (2007, 6, 12))
1

Month day tests
-------------------

>>> inPeriod ("md {12}", datetime (2007, 6, 12))
1

>>> inPeriod ("md {13}", datetime (2007, 6, 12))
0

>>> inPeriod ("md {10-15}", datetime (2007, 6, 12))
1

>>> inPeriod ("md {10 12 14}", datetime (2007, 6, 12))
1

>>> inPeriod ("md {29-14}", datetime (2007, 6, 12))
1

>>> inPeriod ("md {32}", datetime (2007, 6, 12))
Traceback (most recent call last):
InvalidFormat: 32 is not valid for day. Valid options are between 1 and 31.

>>> inPeriod ("md {0}", datetime (2007, 6, 12))
Traceback (most recent call last):
InvalidFormat: 0 is not valid for day. Valid options are between 1 and 31.

>>> inPeriod ("md {tom}", datetime (2007, 6, 12))
Traceback (most recent call last):
InvalidFormat: An integer value is required for day.

>>> inPeriod ("mday {12}", datetime (2007, 6, 12))
1

Weekday tests
-------------------

Match Mondays
>>> inPeriod ("wd {mon}", datetime (2007, 06, 11))
1

Match monday through friday, inclusively
>>> inPeriod ("wd {mon-fri}", datetime (2007, 06, 15))
1

Friday through tuesday
>>> inPeriod ("wd {fri-tue}", datetime (2007, 06, 11))
1

False test
>>> inPeriod ("wd {mon}", datetime (2007, 06, 12))
0

>>> inPeriod ("wd {mon-fri}", datetime (2007, 06, 16))
0

>>> inPeriod ("wd {fri-mon}", datetime (2007, 06, 12))
0

Only the first two characters matter in string names
>>> inPeriod ("wd {tuednesday}", datetime (2007, 06, 12))
1

>>> inPeriod ("wd {june}", datetime (2007, 06, 12))
Traceback (most recent call last):
InvalidFormat: An integer value is required for weekday.

>>> inPeriod ("wday {tue}", datetime (2007, 6, 12))
1

Hour tests
-------------------

>>> inPeriod ("hr {12}", datetime (2007, 6, 12, 12, 05))
1

>>> inPeriod ("hour {12}", datetime (2007, 6, 12, 12))
1

>>> inPeriod ("hr {10-14}", datetime (2007, 6, 12, 12, 05))
1

>>> inPeriod ("hr {10 12 13}", datetime (2007, 6, 12, 12, 05))
1

>>> inPeriod ("hr {12noon}", datetime (2007, 6, 12, 12, 05))
1

>>> inPeriod ("hr {noon}", datetime (2007, 6, 12, 12, 05))
1

>>> inPeriod("hr {12pm}", datetime(2007, 6, 12, 12, 05))
1

>>> inPeriod ("hr {12midnight}", datetime (2007, 6, 12, 0, 05))
1

>>> inPeriod ("hr {midnight}", datetime (2007, 6, 12, 0, 05))
1

>>> inPeriod ("hr {12am}", datetime (2007, 6, 12, 0, 05))
1

>>> inPeriod ("hr {10am-1pm}", datetime (2007, 6, 12, 13, 05))
1

>>> inPeriod ("hr {9pm-5am}", datetime (2007, 6, 12, 9))
0

>>> inPeriod ("hr {13am}", datetime (2007, 6, 12, 12, 05))
Traceback (most recent call last):
InvalidFormat: 13am is an invalid value for hour.

>>> inPeriod ("hr {24}", datetime (2007, 6, 12, 12, 05))
Traceback (most recent call last):
InvalidFormat: 24 is not valid for hour. Valid options are between 0 and 23.

>>> inPeriod ("hr {-1}", datetime (2007, 6, 12, 12, 05))
Traceback (most recent call last):
InvalidFormat: An integer value is required for hour.

>>> inPeriod ("hr {13pm}", datetime (2007, 6, 12, 12, 05))
Traceback (most recent call last):
InvalidFormat: 25 is not valid for hour. Valid options are between 0 and 23.

>>> inPeriod ("hr {fred-george}", datetime (2007, 6, 12, 12, 05))
Traceback (most recent call last):
InvalidFormat: An integer value is required for hour.

Minute Tests
-------------------

>>> inPeriod ("min {5}", datetime (2007, 6, 12, 12, 5))
1

>>> inPeriod ("min {2 4 6}", datetime (2007, 6, 12, 12, 6))
1

>>> inPeriod ("min {0-5}", datetime (2007, 6, 12, 12, 5))
1

>>> inPeriod ("min {0-5}", datetime (2007, 6, 12, 12, 45))
0

>>> inPeriod ("minute {5}", datetime (2007, 6, 12, 12, 5))
1

>>> inPeriod ("min {-1}", datetime (2007, 6, 12, 12, 5))
Traceback (most recent call last):
InvalidFormat: An integer value is required for minute.

>>> inPeriod ("min {60}", datetime (2007, 6, 12, 12, 5))
Traceback (most recent call last):
InvalidFormat: 60 is not valid for minute. Valid options are between 0 and 59.

>>> inPeriod ("min {harry}", datetime (2007, 6, 12, 12, 5))
Traceback (most recent call last):
InvalidFormat: An integer value is required for minute.

Second Tests
-------------------

>>> inPeriod ("sec {5}", datetime (2007, 6, 12, 12, 5, 5))
1

>>> inPeriod ("sec {2 4 6}", datetime (2007, 6, 12, 12, 6, 6))
1

>>> inPeriod ("sec {0-5}", datetime (2007, 6, 12, 12, 5, 5))
1

>>> inPeriod ("sec {0-5}", datetime (2007, 6, 12, 12, 45, 45))
0

>>> inPeriod ("second {5}", datetime (2007, 6, 12, 12, 5, 5))
1

>>> inPeriod ("sec {-1}", datetime (2007, 6, 12, 12, 5))
Traceback (most recent call last):
InvalidFormat: An integer value is required for second.

>>> inPeriod ("sec {60}", datetime (2007, 6, 12, 12, 5))
Traceback (most recent call last):
InvalidFormat: 60 is not valid for second. Valid options are between 0 and 59.

>>> inPeriod ("sec {tom}", datetime (2007, 6, 12, 12, 5))
Traceback (most recent call last):
InvalidFormat: An integer value is required for second.

Time::Period's Examples
-------------------

>>> inPeriod ("wd {Mon-Fri} hr {9am-4pm}", datetime (2007, 8, 1, 12))
1

>>> inPeriod ("wd {Mon Wed Fri} hr {9am-4pm}, wd{Tue Thu} hr {9am-2pm}", datetime (2007, 7, 3, 15))
0

>>> inPeriod ("wk {1 3 5} wd {Mon Wed Fri} hr {9am-4pm}", datetime (2007, 8, 1, 12))
1

>>> inPeriod ("mo {Nov-Feb}", datetime (2007, 1, 1, 12))
1

>>> inPeriod ("mo {Jan-Feb Nov-Dec}", datetime (2007, 8, 1, 12))
0

>>> inPeriod ("mo {jan feb nov dec}", datetime (2007, 11, 1, 12))
1

>>> inPeriod ("mo {Jan Feb}, mo {Nov Dec}", datetime (2007, 12, 1, 12))
1

>>> inPeriod ("mo {Jan Feb} mo {Nov Dec}", datetime (2007, 1, 1, 12))
1

>>> inPeriod ("minute { 0-29 }", datetime (2007, 8, 1, 12, 15))
1

>>> inPeriod ("hour { 12am-11am }", datetime (2007, 8, 1, 1))
1

>>> inPeriod ("sec {0-4 10-14 20-24 30-34 40-44 50-54}", datetime (2007, 8, 1, 12, 15, 22))
1

>>> inPeriod ("wd {1 3 5 7} min {0-29}, wd {2 4 6} min {30-59}", datetime (2007, 6, 10, 12, 45))
0

Empty Ranges
------------

>>> inPeriod('yr {}', datetime(2011, 1, 1))
0

>>> inPeriod('mo {}', datetime(2011, 1, 1))
0

>>> inPeriod('wk {}', datetime(2011, 1, 1))
0

>>> inPeriod('yd {}', datetime(2011, 1, 1))
0

>>> inPeriod('md {}', datetime(2011, 1, 1))
0

>>> inPeriod('wd {}', datetime(2011, 1, 1))
0

>>> inPeriod('hr {}', datetime(2011, 1, 1))
0

>>> inPeriod('min {}', datetime(2011, 1, 1))
0

>>> inPeriod('sec {}', datetime(2011, 1, 1))
0

"""

from TimePeriod import inPeriod
from datetime import datetime

if __name__ == "__main__":
	import doctest
	doctest.testmod ()


TimePeriod.py
"""
Determines if a datetime is within a given time period. This is a Python
version of Perl's Time::Period module.
"""
from datetime import datetime
import calendar
import re

__all__ = ['inPeriod', 'InvalidFormat']

DAYS = {'su': 1,
    'mo': 2,
    'tu': 3,
    'we': 4,
    'th': 5,
    'fr': 6,
    'sa': 7}

MONTHS = {'jan': 1,
    'feb': 2,
    'mar': 3,
    'apr': 4,
    'may': 5,
    'jun': 6,
    'jul': 7,
    'aug': 8,
    'sep': 9,
    'oct': 10,
    'nov': 11,
    'dec': 12}

def inPeriod (period, dt=None):
    """
    Determines if a datetime is within a certain time period. If the time
    is omitted the current time will be used.

    inPeriod returns 1 if the datetime is within the time period, 0 if not.
    If the expression is malformed a TimePeriod.InvalidFormat exception
    will be raised. (Note that this differs from Time::Period, which
    returns -1 if the expression is invalid).

    The format for the time period is like Perl's Time::Period module,
    which is documented in some detail here:

    http://search.cpan.org/~pryan/Period-1.20/Period.pm

    Here's the quick and dirty version.

    Each period is composed of one or more sub-period seperated by a comma.
    A datetime must match at least one of the sub periods to be considered
    in that time period.

    Each sub-period is composed of one or more tests, like so:

        scale {value}

        scale {a-b}

        scale {a b c}

    The datetime must pass each test for a sub-period for the sub-period to
    be considered true.

    For example:

        Match Mondays
        wd {mon}

        Match Monday mornings
        wd {mon} hr {midnight-noon}

        Match Monday morning or Friday afternoon
        wd {mon} hr {midnight-noon}, wd {fri} hr {noon-midnight}

    Valid scales are:
        year
        month
        week
        yday 
        mday
        wday
        hour
        minute
        second

    Those can be substituted with their corresponding code:
        yd
        mo
        wk
        yd
        md
        wd
        hr
        min
        sec
    """
    if dt == None:
        dt = datetime.now ()

    # transform whatever crazy format we're given and turn it into
    # something like this:
    # 
    # md{1}|hr{midnight-noon},md{2}|hr{noon-midnight}
    period = re.sub (r"^\s*|\s*$", "", period)
    period = re.sub (r"\s*(?={|$)", "", period)
    period = re.sub (r",\s*", ",", period)
    period = re.sub (r"\s*-\s*", "-", period)
    period = re.sub (r"{\s*", "{", period)
    period = re.sub (r"\s*}\s*", "}", period)
    period = re.sub (r"}(?=[^,])", "}|", period)
    period = period.lower ()

    if period == "":
        return 1

    sub_periods = re.split (",", period)

    # go through each sub-period until one matches (OR logic)
    for sp in sub_periods:
        if _is_in_sub_period (sp, dt):
            return 1

    return 0

def _is_in_sub_period (sp, dt):
    if sp == "never" or sp == "none":
        return 0
    if sp == "always" or sp == "":
        return 1
    scales = sp.split ("|")
    range_lists = {}

    # build a list for every scale of ranges
    for scale_exp in scales:
        scale, ranges = _parse_scale (scale_exp)

        # if there's already a list for this scale, add the new ranges to the end
        if scale in range_lists:
            range_lists[scale] += ranges
        else:
            range_lists[scale] = ranges

    # check each scale, if there's a false one return false (AND logic)
    for scale in range_lists:
        result = SCALES[scale] (range_lists[scale], dt)
        if result != 1:
            return result

    return 1

def _parse_scale (scale_exp):
    """Parses a scale expression and returns the scale, and a list of ranges."""

    m = re.search ("(\w+?)\{(.*?)\}", scale_exp)
    if m == None:
        raise InvalidFormat ("Unable to parse the given time period.")
    scale = m.group (1)
    range = m.group (2)

    if not scale in SCALES:
        raise InvalidFormat ("%s is not a valid scale." % (scale))

    ranges = re.split ("\s", range)

    return scale, ranges

def yr (ranges, dt):
    def normal_year (year, dt):
        if year == None:
            return year

        try:
            year = int (year)
        except ValueError:
            raise InvalidFormat ("An integer value is required for year.")

        if year < 100:
            century = dt.year / 100
            year = (100 * century)+year

        return year

    for range in ranges:
        low, high = _splitrange (range)
        low = normal_year (low, dt)
        high = normal_year (high, dt)
        
        # reverse the numbers if the first number is higher than the second
        # (e.g. 2015-2010)
        if high != None and low > high:
            tmp = low
            low = high
            high = tmp

        if _is_in_range (dt.year, low, high):
            return 1

    return 0

def mo (ranges, dt):
    for range in ranges:
        for month in MONTHS.keys ():
            range = re.sub ("%s.*?(?=\s|-|$)" % (month), str (MONTHS[month]), range)

        low, high = _splitrange (range)
        low, high = _in_min_max (low, high, 1, 12, "month")

        if _is_in_range (dt.month, low, high):
            return 1

    return 0

def wk (ranges, dt):
    # Calculate the week number. This is essentially the number of Sunday's
    # that have passed.
    week = 1
    for day in range(1, dt.day + 1):
        if (day != 1 and calendar.weekday(dt.year, dt.month, day) == 6):
            week += 1

    return _simple_test (week, ranges, 1, 6, "week")

def yd (ranges, dt):
    today = int (dt.strftime ("%j"))
    return _simple_test (today, ranges, 1, 366, "year day")

def md (ranges, dt):
    return _simple_test (dt.day, ranges, 1, 31, "day")

def wd (ranges, dt):
    today = (dt.weekday () + 2) % 7
    # saturday needs to be 7 not zero
    if today == 0:
        today = 7

    for range in ranges:
        # translate days into numbers
        for day in DAYS.keys ():
            range = re.sub ("%s.*?(?=\s|-|$)" % (day), str (DAYS[day]), range)

        low, high = _splitrange (range)
        low, high = _in_min_max (low, high, 1, 7, "weekday")

        if _is_in_range (today, low, high):
            return 1
    return 0

def hr (ranges, dt):
    now = dt.hour

    def normal_hour (hour):
        if hour == None:
            return None
        try:
            if hour.endswith ("am"):
                hour = int (hour[:-2])
                if hour > 12:
                    # things like 13am would actually work, but since it's
                    # invalid we're clinging to our sensibilities and throwing
                    # it out
                    raise InvalidFormat ("%sam is an invalid value for hour." % hour)
                elif hour == 12:    # 12am is midnight
                    hour = 0
            elif hour.endswith ("pm"):
                hour = int (hour[:-2])
                if hour != 12:      # 12pm is noon
                    hour += 12
            elif hour == "12noon" or hour == "noon":
                hour = 12
            elif hour == "12midnight" or hour == "midnight":
                hour = 0
            else:
                hour = int (hour)
        except ValueError:
            raise InvalidFormat ("An integer value is required for hour.")

        return hour

    for range in ranges:
        low, high = _splitrange (range)    
        low = normal_hour (low)
        high = normal_hour (high)

        low, high = _in_min_max (low, high, 0, 23, "hour")
        if _is_in_range (now, low, high):
            return 1

    return 0

def min (ranges, dt):
    return _simple_test (dt.minute, ranges, 0, 59, "minute")

def sec (ranges, dt):
    return _simple_test (dt.second, ranges, 0, 59, "second")

def _simple_test (now, ranges, min, max, scale):
    for range in ranges:
        low, high = _splitrange (range)
        low, high = _in_min_max (low, high, min, max, scale)

        if _is_in_range (now, low, high):
            return 1

    return 0

def _splitrange (range):
    if range == "":
        return None, None

    lowhigh = range.split ("-")
    low = lowhigh[0]
    high = None
    if len(lowhigh) > 1:
        high = lowhigh[1]

    return low, high

def _in_min_max (low, high, min, max, scale):
    if low == None:
        return low, high

    try:
        low = int(low)
    except ValueError:
        raise InvalidFormat ("An integer value is required for %s." % scale)

    if low < min or low > max:
        raise InvalidFormat ("%d is not valid for %s. Valid options are between %d and %d." % (low, scale, min, max))
    if high != None:
        try:
            high = int (high)
        except ValueError:
            raise InvalidFormat ("An integer value is required for %s." % scale)

        if high < min or high > max:
            raise InvalidFormat ("%d is not valid for %s. Valid options are between %d and %d." % (low, scale, min, max))

    return low, high

def _is_in_range (x, low, high):
    if low == None and high == None:
        # empty range, never matches
        return 0

    if high == None or low == high:
        # just one number
        if low != x:
            return 0
    elif high > low: # e.g. mon-fri
        if x > high or x < low:
            return 0
    elif low > high: # e.g. fri-mon
        if not (x >= low or x <= high):
            return 0

    return 1

class InvalidFormat (Exception):
    pass

# dict of scale codes with the functions to process them
SCALES = {
    'yr': yr, 'year': yr,
    'mo': mo, 'month': mo,
    'wk': wk, 'week': wk,
    'yd': yd, 'yday': yd,
    'md': md, 'mday': md,
    'wd': wd, 'wday': wd,
    'hr': hr, 'hour': hr,
    'min': min, 'minute': min,
    'sec': sec, 'second': sec,
    }

if __name__ == "__main__":
    import sys
    try:
        if len(sys.argv) == 3:
            epoch = int(sys.argv[1])
            period = sys.argv[2]
            print inPeriod (period, datetime.fromtimestamp(epoch))
    except InvalidFormat, ife:
        print "Invalid time period format: %s" % ife
